test/integration/graphql_test.lua

```diff
@@ -687,6 +734,105 @@ g.test_custom_type_scalar_variables = function()
     end)
 end
 
+g.test_output_type_mismatch_error = function()
+    local server = cluster.main_server
+
+    server.net_box:eval([[
+        package.loaded['test'] = {}
+        package.loaded['test']['callback'] = function(_, _)
+            return true
+        end
+
+        package.loaded['test']['callback_for_nested'] = function(_, _)
+            return { values = true }
+        end
+
+        local graphql = require('cartridge.graphql')
+        local types = require('cartridge.graphql.types')
+
+        local obj_type = types.object({
+            name = 'ObjectWithValue',
+            fields = {
+                value = types.string,
+            },
+        })
+
+        local nested_obj_type = types.object({
+            name = 'NestedObjectWithValue',
+            fields = {
+                value = types.string,
+            },
+        })
+
+        local complex_obj_type = types.object({
+            name = 'ComplexObjectWithValue',
+            fields = {
+                values = types.list(nested_obj_type),
+            },
+        })
+
+        graphql.add_callback({
+            name = 'expected_nonnull_list',
+            kind = types.list(types.int.nonNull),
+            callback = 'test.callback',
+        })
+
+        graphql.add_callback({
+            name = 'expected_obj',
+            kind = obj_type,
+            callback = 'test.callback',
+        })
+
+        graphql.add_callback({
+            name = 'expected_list',
+            kind = types.list(types.int),
+            callback = 'test.callback',
+        })
+
+        graphql.add_callback({
+            name = 'expected_list_with_nested',
+            kind = types.list(complex_obj_type),
+            callback = 'test.callback_for_nested',
+        })
+    ]])
+
+    t.assert_error_msg_equals('Expected "expected_nonnull_list" to be an "array", got "boolean"', function()
+        return server:graphql({
+            query = [[
+                query {
+                    expected_nonnull_list
+                }
+            ]]})
+    end)
+
+    t.assert_error_msg_equals('Expected "expected_obj" to be a "map", got "boolean"', function()
+        return server:graphql({
+            query = [[
+                query {
+                    expected_obj { value }
+                }
+            ]]})
+    end)
+
+    t.assert_error_msg_equals('Expected "expected_list" to be an "array", got "boolean"', function()
+        return server:graphql({
+            query = [[
+                query {
+                    expected_list
+                }
+            ]]})
+    end)
+
+    t.assert_error_msg_equals('Expected "expected_list_with_nested" to be an "array", got "table"', function()
```

Comment:
When using `array`/`map` notation, "table" become confusing.

Reply:
Do you have any alternatives?

Reply:
What about adding additional check - when `result` is a table, check that it has no string keys via `type(next(result))`. If so, print "map" in the error message. If not, just print result's type.

Reply:
Seems could be simply done with:
```lua
local resultType = type(result)
if resultType == 'table' then
    resultType = 'map'
end
```

Reply:
Finally, `Expected "expected_list_with_nested" to be an "array", got "map"`